### PR TITLE
[ACR] Task GA commands

### PR DESCRIPTION
**Pull request by ankurkhemani**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_help.py [hunk 1]
```diff
@@ -392,143 +397,274 @@
...
+helps['acr task run'] = """
+    type: command
+    short-summary: Queues a quick run providing streamed logs for an Azure Container Registry.
+    examples:
+        - name: Queue a local context, pushed to ACR with streaming logs.
+          text: >
+            az acr task run -r myregistry --task-file bash-echo.yaml .
+        - name: Queue a remote git context with streaming logs.
+          text: >
+            az acr run -r myregistry https://github.com/AzureCR/acr-tasks-sample.git --task-file bash-echo.yaml
```
Comment (ehotinger) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_help.py]: this needs to be removed

File: src/command_modules/azure-cli-acr/setup.py [hunk 2]
```diff
@@ -56,6 +56,9 @@
         'azure.cli',
         'azure.cli.command_modules',
         'azure.cli.command_modules.acr',
+        'azure.cli.command_modules.acr.sdk',
+        'azure.cli.command_modules.acr.sdk.models',
+        'azure.cli.command_modules.acr.sdk.operations'
```
Comment (ehotinger) [inline src/command_modules/azure-cli-acr/setup.py]: TODO: remove when SDK is published

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_help.py (hunk 1 above)
Comment (ankurkhemani) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_help.py]: Done.

File: src/command_modules/azure-cli-acr/setup.py (hunk 2 above)
Comment (ankurkhemani) [reply src/command_modules/azure-cli-acr/setup.py]: Removed.

**Comment by ankurkhemani** [comment]
@tjprescott @troydai We are using version 2.2.0 of the sdk from the "privates" folder and looks like the CI is complaining because it cannot locate the sdk? Can you please look into it?

**Comment by troydai** [comment]
I will take a quick look. By the way, when is the SDK going to be released?

**Comment by troydai** [comment]
You can temporarily disable the dependency check task in the `.travis.yml` file. And re-enable it once the SDK is released. We do not allow private whl enter the `dev` branch on principle.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_stream_utils.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+from io import BytesIO
+import time
+from random import uniform
+import colorama
+from knack.util import CLIError
+from knack.log import get_logger
+from msrestazure.azure_exceptions import CloudError
+from azure.mgmt.containerregistry.v2018_02_01_preview.operations import BuildsOperations
```
[the remark this answers is not in the document]
Comment (ankurkhemani) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_stream_utils.py]: Yes, this is for backward compatibility.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py [hunk 4]
```diff
@@ -0,0 +1,472 @@
...
+                    source_repository=SourceUpdateParameters(
+                        source_control_type=source_control_type,
+                        repository_url=context_path,
+                        branch=branch,
+                        source_control_auth_properties=AuthInfoUpdateParameters(
+                            token=git_access_token,
+                            token_type=DEFAULT_TOKEN_TYPE
+                        )
+                    ),
+                    # TODO: pull request?
```
[the remark this answers is not in the document]
Comment (jaysterp) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py]: We are adding PR as a source trigger event alongside commit. However, it is not being included in this release. Will remove the TODO.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py [hunk 5]
```diff
@@ -0,0 +1,472 @@
...
+                        source_control_type=source_control_type,
+                        repository_url=context_path,
+                        branch=branch,
+                        source_control_auth_properties=AuthInfo(
+                            token=git_access_token,
+                            token_type=DEFAULT_TOKEN_TYPE,
+                            scope='repo'
+                        )
+                    ),
+                    # TODO: pull request?
```
[the remark this answers is not in the document]
Comment (ankurkhemani) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py]: There was a plan for pull request trigger. We are not doing it this GA, will remove.

**Comment by ankurkhemani** [comment]
@troydai Can we please get this PR merged today, this is aimed for Ignite. Thanks!

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_archive_utils.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+from knack.log import get_logger
+from knack.util import CLIError
+from msrestazure.azure_exceptions import CloudError
+from azure.storage.blob import BlockBlobService
+from ._azure_utils import get_blob_info
+
+logger = get_logger(__name__)
+
+
+def upload_source_code(client, registry_name, resource_group_name,
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_archive_utils.py]: NIT: either put all of them in one line or each arg takes one line.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_archive_utils.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+                       source_location, tar_file_path,
+                       docker_file_path, docker_file_in_tar):
+    _pack_source_code(source_location,
+                      tar_file_path,
+                      docker_file_path,
+                      docker_file_in_tar)
+
+    size = os.path.getsize(tar_file_path)
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_archive_utils.py]: Any possible `OSError` here?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_archive_utils.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+        size, unit, registry_name))
+    return relative_path
+
+
+def _pack_source_code(source_location, tar_file_path, docker_file_path, docker_file_in_tar):
+    logger.warning("Packing source code into tar to upload...")
+
+    ignore_list, ignore_list_size = _load_dockerignore_file(source_location)
+    common_vcs_ignore_list = {'.git', '.gitignore',
+                              '.bzr', 'bzrignore', '.hg', '.hgignore', '.svn'}
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_archive_utils.py]: NIT: unnecessary line breaking

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_archive_utils.py [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+
+    def _ignore_check(tarinfo, parent_ignored, parent_matching_rule_index):
+        # ignore common vcs dir or file
+        if tarinfo.name in common_vcs_ignore_list:
+            logger.warning(
+                "Excluding '%s' based on default ignore rules", tarinfo.name)
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_archive_utils.py]: NIT: unnecessary line breaking

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_azure_utils.py [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+                       r"(?P<blob_name>.*?)\?(?P<sas_token>.*)"), blob_sas_url)
+    account_name = match.group('account_name')
+    endpoint_suffix = match.group('endpoint_suffix')
+    container_name = match.group('container_name')
+    blob_name = match.group('blob_name')
+    sas_token = match.group('sas_token')
+
+    if not account_name or not container_name or not blob_name or not sas_token:
+        raise CLIError(
+            "Failed to parse the SAS URL: '{!s}'.".format(blob_sas_url))
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_azure_utils.py]: NIT: unnecessary line breaking

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_stream_utils.py [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+    # didn't end in \r\n, so we were unable to flush out the final contents.
+    curr_bytes = stream.getvalue()
+    if curr_bytes:
+        print(curr_bytes.decode('utf-8', errors='ignore'))
+
+    build_status = _get_run_status(metadata).lower()
+    logger.debug("status was: '%s'", build_status)
+
+    if raise_error_on_failure:
+        if build_status == 'internalerror' or build_status == 'failed':
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_stream_utils.py]: Do we have enum strings for `build_status`?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py [hunk 12]
```diff
@@ -3,288 +3,93 @@
...
+              docker_file_path='',
               no_format=False,
               no_push=False,
               no_logs=False,
-              os_type='Linux'):
+              os_type=OS.linux.value):
     _, resource_group_name = validate_managed_registry(
         cmd.cli_ctx, registry_name, resource_group_name, BUILD_NOT_SUPPORTED)
 
+    from ._client_factory import cf_acr_registries
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: Does the new API support both quick build and quick run?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py [hunk 13]
```diff
@@ -293,210 +98,33 @@ def acr_build(cmd,
...
         registry_name=registry_name,
-        build_request=build_request))
+        run_request=docker_build_request))
 
-    build_id = queued_build.build_id
-    logger.warning("Queued a build with build ID: %s", build_id)
-    logger.warning("Waiting for build agent...")
+    run_id = queued_build.run_id
+    logger.warning("Queued a build with ID: %s", run_id)
+    logger.warning("Waiting for agent...")
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: Do we want to support no-wait?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py [hunk 14]
```diff
@@ -0,0 +1,469 @@
...
+            docker_file_path=file,
+            arguments=(arg if arg else []) + (secret_arg if secret_arg else []),
+            context_path=context_path
+        )
+
+    registry, resource_group_name = validate_managed_registry(
+        cmd.cli_ctx, registry_name, resource_group_name, TASK_NOT_SUPPORTED)
+
+    source_control_type = SourceControlType.visual_studio_team_service.value
+    if context_path is not None and 'GITHUB.COM' in context_path.upper():
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py]: NIT: Any reason why sometimes we check for upper() and sometimes check for lower()?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py [hunk 15]
```diff
@@ -0,0 +1,469 @@
...
+            os=os_type,
+            architecture=Architecture.amd64.value
+        ),
+        status=status,
+        timeout=timeout,
+        agent_configuration=AgentProperties(
+            cpu=cpu
+        ),
+        trigger=TriggerProperties(
+            source_triggers=[
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py]: If `source_triggers` is an array, how do we support adding multiple of them?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py [hunk 16]
```diff
@@ -0,0 +1,469 @@
...
+                        branch=branch,
+                        source_control_auth_properties=AuthInfo(
+                            token=git_access_token,
+                            token_type=DEFAULT_TOKEN_TYPE,
+                            scope='repo'
+                        )
+                    ),
+                    source_trigger_events=[SourceTriggerEvent.commit.value],
+                    status=TriggerStatus.enabled.value if commit_trigger_enabled else TriggerStatus.disabled.value,
+                    name=source_trigger_name if source_trigger_name else "defaultSourceTriggerName"
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py]: Make `defaultSourceTriggerName` a const?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py [hunk 17]
```diff
@@ -0,0 +1,469 @@
...
+                )
+            ] if commit_trigger_enabled else None,
+            base_image_trigger=BaseImageTrigger(
+                base_image_trigger_type=base_image_trigger_type,
+                status=TriggerStatus.enabled.value if base_image_trigger_enabled else TriggerStatus.disabled.value,
+                name=base_image_trigger_name if base_image_trigger_name else "defaultBaseimageTriggerName"
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py]: Make `defaultBaseimageTriggerName` a const?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py [hunk 18]
```diff
@@ -0,0 +1,469 @@
...
+                            token=git_access_token,
+                            token_type=DEFAULT_TOKEN_TYPE,
+                            scope='repo'
+                        )
+                    ),
+                    source_trigger_events=[SourceTriggerEvent.commit.value],
+                    status=TriggerStatus.enabled.value if commit_trigger_enabled else TriggerStatus.disabled.value,
+                    name=source_trigger_name if source_trigger_name else "defaultSourceTriggerName"
+                )
+            ] if commit_trigger_enabled else None,
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py]: I felt this is hard to read... Can we do something like
```
source_trigger = None
if commit_trigger_enabled:
  source_trigger = <10+ line block>
# assign source_trigger to task_create_parameters
```

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py [hunk 19]
```diff
@@ -0,0 +1,469 @@
...
+            base_image_trigger=BaseImageTrigger(
+                base_image_trigger_type=base_image_trigger_type,
+                status=TriggerStatus.enabled.value if base_image_trigger_enabled else TriggerStatus.disabled.value,
+                name=base_image_trigger_name if base_image_trigger_name else "defaultBaseimageTriggerName"
+            ) if base_image_trigger_enabled else None
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py]: Same as `commit_trigger_enabled`

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py [hunk 20]
```diff
@@ -0,0 +1,469 @@
...
+                    secret_arg=None,
+                    set_value=None,
+                    set_secret=None,
+                    base_image_trigger_enabled=None,
+                    base_image_trigger_type=None):
+    _, resource_group_name = validate_managed_registry(
+        cmd.cli_ctx, registry_name, resource_group_name, TASK_NOT_SUPPORTED)
+
+    task = client.get(resource_group_name, registry_name, task_name)
+    step = task.step
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py]: Does `step` always exist in `task`?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py [hunk 21]
```diff
@@ -0,0 +1,469 @@
...
+    if isinstance(step, DockerBuildStep):
+        if file and file.endswith(ALLOWED_TASK_FILE_TYPES):
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py]: Why do we throw if file ends with allowed file types?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py [hunk 22]
```diff
@@ -0,0 +1,469 @@
...
+        )
+    elif isinstance(step, FileTaskStep):
+        if file and not file.endswith(ALLOWED_TASK_FILE_TYPES):
+            raise CLIError("File for task build step has an invalid suffix: {}."
+                           " It must have one of the following suffixes: {}"
+                           .format(file, ALLOWED_TASK_FILE_TYPES))
+        if set_value is None and set_secret is None:
+            values = None
+        else:
+            values = (set_value if set_value else []) + (set_secret if set_secret else [])
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py]: NIT: this if-else block can be moved out of the outer if-else so it won't be duplicated.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py [hunk 23]
```diff
@@ -0,0 +1,469 @@
...
+        step = FileTaskStepUpdateParameters(
+            task_file_path=file,
+            values_file_path=values,
+            context_path=context_path,
+            values=values
+        )
+
+    source_control_type = None
+    if context_path is not None:
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py]: if context_path:

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py [hunk 24]
```diff
@@ -0,0 +1,469 @@
...
+            repository, _, manifest = get_image_digest(cmd.cli_ctx, registry_name, resource_group_name, image)
+            filter_str = _add_run_filter(
+                filter_str, 'OutputImageManifests', '{}@{}'.format(repository, manifest), 'contains')
+        except CLIError as e:
+            raise CLIError("Could not find image '{}'. {}".format(image, e))
+
+    return client.list(resource_group_name, registry_name, filter=filter_str, top=top)
+
+
+def _add_run_filter(orig_filter, name, value, operator):
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py]: NIT: quite some duplicated code from build_task.py

Comment (ankurkhemani) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py]: I would keep this, and nuke the `build_task.py` file in the next iteration when we deprecate `build-task`. Don't want to create a `_build_task_utils.py`

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py (hunk 21 above)
Comment (ankurkhemani) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py]: Earlier, we didn't want to allow updating step to a different type than what was created. That changed today morning, have pushed changes.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py (hunk 15 above)
Comment (ankurkhemani) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/task.py]: We don't currently.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 12 above)
Comment (ankurkhemani) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: No. For `az acr build` we schedule a run and hence using the new api.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py (hunk 13 above)
Comment (ankurkhemani) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: Not here. // cc @mnltejaswini

**Comment by sajayantony** [comment]
LGTM.

**Comment by troydai** [comment]
Waiting for CI

**Comment by ankurkhemani** [comment]
Ready to merge if CI passes.Thanks!

**Comment by sajayantony** [comment]
Thanks for the last minute help fixes @ankurkhemani

**Comment by sajayantony** [comment]
LGTM.

**Pull request merged**